Install script fixes for better Mac support

HudsonOnHere opened the pull request:
Trying to install on Mac and encountered a bug. At line 18, sha256sum was throwing an error because that command has a different syntax on Macs.

_Verifiying checksum...
Error: ./support/scripts/install: line 18: sha256sum: command not found_

Fixed this by adding additional logic for checksum validation if `uname -s` returns "Darwin"

Created an array for supported system types, for better readability and future extensibility.

sundowndev reviewed (changes requested):
Thanks for your PR @HudsonOnHere 👍🏻 

However I don't think that's a great way of handling this. Verifying checksum is optional and sha256sum could be missing on other platforms too (assuming this script can evolve at any time to support more platforms). Instead, I would suggest we simply skip checksum verification and display a warning message such as "WARNING: checksum verification was skipped because sha256sum is not installed."

WDYT?

sundowndev commented:
Eventually, this PR could fix https://github.com/sundowndev/phoneinfoga/issues/1202 as well. I think we can drop usage of wget and use curl for all requests.

HudsonOnHere commented:
> Thanks for your PR @HudsonOnHere 👍🏻
> 
> However I don't think that's a great way of handling this. Verifying checksum is optional and sha256sum could be missing on other platforms too (assuming this script can evolve at any time to support more platforms). Instead, I would suggest we simply skip checksum verification and display a warning message such as "WARNING: checksum verification was skipped because sha256sum is not installed."
> 
> EDIT: another solution could be to check whever sha256sum or shasum is installed and use one of them, without checking the platform. If none is installed, exit with error.
> 
> WDYT?

@sundowndev Thanks for your feedback -- I like your idea about checking for which shasum is installed, I did not realize that the intention was for this validation to be optional, so I will update the code so that it does not fail if the validation does not succeed.


> Eventually, this PR could fix #1202 as well. I think we can drop usage of wget and use curl for all requests.

I can also update the code to use `curl` here instead, that should be a simple fix.

sundowndev commented:
@HudsonOnHere Thanks for following up!

On a second thought, let's perform the check on which program is available and then still exit with error when the verification fails. However, we should display a warning message if neither sha256sum or shasum is installed.

HudsonOnHere commented:
Sounds good to me -- I should have some updates in the next day or two.

HudsonOnHere commented:
@sundowndev I think this update should address all of our concerns. Sorry for the amount of changes introduced, but I thought it would be best to refactor the code so that we could handle everything better and make it more robust.

Here's what changed:
- Made checksum validation mandatory
- added an optional flag to skip checksum in case it's needed
- added a manual download mode as an option, or as a fallback in case the automatic check with `uname` fails for some reason.
- added a detailed help message with usage examples
- added checks for `wget` before downloading, if not installed use `curl`
- added checks for `sha256sum` before checksum validation, if not installedl use `shasum`

HudsonOnHere commented on support/scripts/install:
```diff
@@ -3,67 +3,219 @@
+
+validate_checksum() {
+
+    local WGET_INSTALLED=false
+    which wget > /dev/null
+    [ $? -eq 1 ] || WGET_INSTALLED=true
+
+    echo "Validating checksum ..."
+    if $WGET_INSTALLED; then
+        wget --quiet --show-progress "https://github.com/sundowndev/phoneinfoga/releases/download/$PHONEINFOGA_VERSION/phoneinfoga_checksums.txt"
+    elif ! $WGET_INSTALLED; then
+        curl --progress-bar -LOC - "https://github.com/sundowndev/phoneinfoga/releases/download/$PHONEINFOGA_VERSION/phoneinfoga_checksums.txt"
+    fi
```
It's more of a "just in case" thing, if for whatever reason some odd linux distro comes with wget installed instead of curl it won't fail.

sundowndev replied:
But it already uses curl everywhere else so I guess the script would fail earlier if curl wouldn't be installed

HudsonOnHere replied:
curl is now the primary downloader, but I left in secondary checks for wget for any "just in case" scenarios. I think it's safer to have the option to fall back on it if there's ever a need.

sundowndev replied:
I get your point but I just don't want to maintain the wget dependency here, I'd rather keep this script simple and easy to maintain. If the user doesn't have curl, tar, and shasum/sha256sum, then the script cannot run.

HudsonOnHere commented on support/scripts/install:
```diff
@@ -3,67 +3,219 @@
+        echo "ERROR: Unable to determine your system type."
         exit 1
     fi
+}
```
Oops this is supposed to be used at the beginning of the main function below, let me take another look at it

HudsonOnHere replied:
This is now corrected -- Added back to main() as was originally intended

HudsonOnHere commented on support/scripts/install:
```diff
@@ -3,67 +3,219 @@
+    [ $? -eq 1 ] || WGET_INSTALLED=true
+
+    echo "Please select the version of PhoneInfoga that you need."
+    PS3="Enter a number: "
```
What system are you using? It was working on my Mac. I'll do some more testing on my Arch machine before updating

HudsonOnHere replied:
When you get the chance, can you please give it another try? I think these updates should resolve this bug.

sundowndev replied:
Still doesn't work on my end, I think you should use `read` instead

HudsonOnHere replied:
Sorry one more question, because I'm struggling to find an explanation for the behavior you're seeing. Did you run this code block separately from the rest of the script? I see `curl -sSL https://...` but that is not part of the original code.

Otherwise, can you please explain what you mean by using `read`? Do you mean instead of `select`? Thanks!

HudsonOnHere replied:
I was able to replicate the issue, it is an oddity of `bash` and I was able to get around it using this command to fetch & run the script instead.

#### Suggested solution:
Update the documentation to include this command instead.
```
bash <( curl -sSL https://raw.githubusercontent.com/sundowndev/phoneinfoga/78c4478f7b6988531e041e9196b0ca768cef975e/support/scripts/install )
```
<br>

this solution also works perfectly fine in automatic mode (above example) or when adding single/multiple arguments to the end of the command, like this.
```
bash <( curl -sSL https://raw.githubusercontent.com/sundowndev/phoneinfoga/78c4478f7b6988531e041e9196b0ca768cef975e/support/scripts/install ) -m -s
```

<br>

[This](https://stackoverflow.com/a/4642978) is where I discovered the solution, after a few failed other attempts.

Updating the one-liner in your documentation's [Getting Started](https://sundowndev.github.io/phoneinfoga/getting-started/install/#binary-installation-recommended) should remedy this behavior.

sundowndev replied:
Fair enough, thanks for the research!

sundowndev reviewed (changes requested):
I think we're almost good.

Remaining changes:

- Fix the user input not working on manual mode
- Fix issues highlighted by [shellcheck](https://www.shellcheck.net/) (see below)

HudsonOnHere commented:
> * Fix the user input not working on manual mode
@sundowndev Please see my comment above Re: user input on manual mode -- I need some additional feedback for debugging

> * Fix issues highlighted by [shellcheck](https://www.shellcheck.net/) (see below)
These issues have been corrected

sundowndev commented on support/scripts/install:
```diff
@@ -1,38 +1,259 @@
+        "Linux_x86_64"
+        "Windows_arm64"
+        "Windows_armv6"
+        "Windows_armv7"
+        "Windows_i386"
+        "Windows_x86_64"
```
You may have guessed those targets but they actually don't exist in releases. On a side note, since this script is using jq, I think it'd be cool to fetch that list from the github api so it's always up to date. WDYT?

sundowndev reviewed (changes requested):
Left a comment about the list of supported OS

HudsonOnHere commented:
> Left a comment about the list of supported OS

Sorry I must be missing it, I don't see anything

sundowndev commented:
@HudsonOnHere Sorry it was a mistake on my end.
